akka-actor/src/main/scala-2.13+/akka/util/ByteString.scala

```diff
@@ -361,10 +367,22 @@ object ByteString {
     def asByteBuffers: scala.collection.immutable.Iterable[ByteBuffer] = List(asByteBuffer)
 
     override def decodeString(charset: String): String =
-      new String(if (length == bytes.length) bytes else toArray, charset)
+      if (isEmpty) ""
+      else new String(bytes, startIndex, length, charset)
 
     override def decodeString(charset: Charset): String = // avoids Charset.forName lookup in String internals
-      new String(if (length == bytes.length) bytes else toArray, charset)
+      if (isEmpty) ""
+      else new String(bytes, startIndex, length, charset)
+
+    override def decodeBase64: ByteString =
+      if (isEmpty) this
+      else if (startIndex == 0 && length == bytes.length) ByteString1C(Base64.getDecoder.decode(bytes))
+      else ByteString1C(Base64.getDecoder.decode(ByteBuffer.wrap(bytes, startIndex, length)).array())
```

Comment:
It's unfortunate that we need to implement this as the above, because Decoder.decode doesn't have an overload which accepts startIndex and length. So we need to use the ByteBuffer overload to at least only allocate a wrapper rather than allocating and throwing away an array.

Reply:
Is the ByteBuffer that's returned by Base64 guaranteed to be array backed? Perhaps a check is needed just in case it changes in a future JDK version? Also, is the array guaranteed to be the same length as the base64ed bytes? Or is it possible that it may return an array that's bigger?

Reply:
@jroper I think that's a good question. I'll introduce a fallback in the case that some platform decides to return native ByteBuffers.